src/showingpreviously/cinemas/cineworld.py

```diff
@@ -0,0 +1,105 @@
+
+
+CINEMAS_API_URL = 'https://www.cineworld.co.uk/uk/data-api-service/v1/quickbook/10108/cinemas/with-event/until/{end_date}'
+SHOWINGS_API_URL = 'https://www.cineworld.co.uk/uk/data-api-service/v1/quickbook/10108/film-events/in-cinema/{cinema_id}/at-date/{date}'
+
+CHAIN = Chain('Cineworld UK')
+
+
+def get_response(url: str) -> requests.Response:
+    r = requests.get(url)
+    if r.status_code != 200:
+        raise CinemaArchiverException(f'Got status code {r.status_code} when fetching URL {url}')
+    return r
+
+
+def get_cinemas_as_dict() -> dict[str, Cinema]:
+    end_date = datetime.now() + timedelta(days=STANDARD_DAYS_AHEAD)
+    url = CINEMAS_API_URL.format(end_date=end_date.strftime('%Y-%m-%d'))
+    r = get_response(url)
+    try:
+        cinemas_data = r.json()
+    except json.JSONDecodeError:
+        raise CinemaArchiverException(f'Error decoding JSON data from URL {url}')
+    cinemas = {}
+    for cinema in cinemas_data['body']['cinemas']:
+        id = cinema['id']
+        name = cinema['displayName']
+        timezone = UK_TIMEZONE
+        cinemas[id] = Cinema(name, timezone)
+    return cinemas
+
+
+def get_json_attributes(attributes: [str]) -> dict[str, any]:
+    json_attributes = {}
+    for attribute in attributes:
+        if attribute == 'audio-described':
+            json_attributes['audio-described'] = True
+        elif attribute in ['120-fps', '35-mm', '70-mm', '4dx', '4k', 'imax', 'imax-3d', 'imax-3d-vip', 'imax-vr', 'screenx', 'thx', '3d']:
+            if 'format' not in json_attributes:
+                json_attributes['format'] = []
+            json_attributes['format'].append(attribute)
+        elif attribute in ['bengali', 'chinese-st', 'filipino', 'gujarati', 'hindi', 'japanese', 'kannada', 'korean', 'malayalam', 'mandarin', 'marathi', 'nepali', 'punjabi', 'russian', 'spanish', 'tamil', 'telugu', 'urdu', 'vietnamese']:
+            json_attributes['language'] = attribute
+        elif attribute in ['sub-titled', 'subbed', 'korean-sub', 'spanish-sub']:
+            if attribute == 'korean-sub':
+                json_attributes['subtitled'] = 'korean'
+            elif attribute == 'spanish-sub':
+                json_attributes['subtitled'] = 'spanish'
+            else:
+                json_attributes['subtitled'] = True
+        elif attribute in ['dubbed', 'eng-dubbed', 'spanish-dub']:
+            # these should already be handled by the language the film is in
+            pass
+
+    return json_attributes
+
+
+def get_showings_date(cinema_id: str, cinema: Cinema, date: str) -> [Showing]:
+    url = SHOWINGS_API_URL.format(cinema_id=cinema_id, date=date)
+    r = get_response(url)
+    try:
+        showings_data = r.json()
+    except json.JSONDecodeError:
+        raise CinemaArchiverException(f'Error decoding JSON data from URL {url}')
+
+    films = {}
+    for film in showings_data['body']['films']:
+        id = film['id']
+        name = film['name']
+        year = film['releaseYear']
```

Comment:
Be aware that this is not 100% accurate. I found a showing of Dune (1984) at the Castleford Cineworld that was marked as being released in 2021. (Not a problem for Cineworld, since it has named the 2021 remake as "Dune (2021)", but this is going to lead to an incorrect entry in the DB).

Not much we can do about this really..... unless we have a "metadata" field that we can put things like the director, actors, etc... in? I'll raise this as a separate issue.

Reply:
Yes, this is an issue with other cinemas too. Like the Harry Potter 20th anniversary screening release date is often listed as "2021" - which I guess is *technically* correct for the 20th anniversary screening, but isn't exactly what we want. Director is an interesting idea, I'll have a think and reply on your issue. For now best to approve these, and we can add director in later to all the archivers.